src/nvim/eval/encode.c

```diff
@@ -0,0 +1,1291 @@
+                  utf_len - (i - shift), utf_buf + i - shift);
+            xfree(tofree);
+            return FAIL;
+          } else if ((SURROGATE_HI_START <= ch && ch <= SURROGATE_HI_END)
+                     || (SURROGATE_LO_START <= ch && ch <= SURROGATE_LO_END)) {
+            emsgf(_("E474: UTF-8 string contains code point which belongs "
+                    "to a surrogate pair: %.*s"),
+                  utf_len - (i - shift), utf_buf + i - shift);
+            xfree(tofree);
+            return FAIL;
+          } else if (ENCODE_RAW(p_enc_conv, ch)) {
+            str_len += shift;
+          } else {
+            str_len += ((sizeof("\\u1234") - 1)
+                        * (size_t) (1 + (ch >= SURROGATE_FIRST_CHAR)));
+          }
+          break;
+        }
+      }
+    }
+    ga_append(gap, '"');
+    ga_grow(gap, (int) str_len);
+    for (size_t i = 0; i < utf_len;) {
+      const int ch = utf_ptr2char(utf_buf + i);
+      const size_t shift = (ch == 0? 1: utf_char2len(ch));
+      assert(shift > 0);
+      // Is false on invalid unicode, but this should already be handled.
+      assert(ch == 0 || shift == utf_ptr2len(utf_buf + i));
+      switch (ch) {
+        case BS:
+        case TAB:
+        case NL:
+        case FF:
+        case CAR:
+        case '"':
+        case '\\': {
+          ga_concat_len(gap, escapes[ch], 2);
+          break;
+        }
+        default: {
+          if (ENCODE_RAW(p_enc_conv, ch)) {
+            ga_concat_len(gap, utf_buf + i, shift);
+          } else if (ch < SURROGATE_FIRST_CHAR) {
+            ga_concat_len(gap, ((const char[]) {
+                '\\', 'u',
+                xdigits[(ch >> (4 * 3)) & 0xF],
+                xdigits[(ch >> (4 * 2)) & 0xF],
+                xdigits[(ch >> (4 * 1)) & 0xF],
+                xdigits[(ch >> (4 * 0)) & 0xF],
+            }), sizeof("\\u1234") - 1);
+          } else {
+            const int tmp = ch - SURROGATE_FIRST_CHAR;
+            const int hi = SURROGATE_HI_START + ((tmp >> 10) & ((1 << 10) - 1));
+            const int lo = SURROGATE_LO_END + ((tmp >>  0) & ((1 << 10) - 1));
+            ga_concat_len(gap, ((const char[]) {
+                '\\', 'u',
+                xdigits[(hi >> (4 * 3)) & 0xF],
+                xdigits[(hi >> (4 * 2)) & 0xF],
+                xdigits[(hi >> (4 * 1)) & 0xF],
+                xdigits[(hi >> (4 * 0)) & 0xF],
+                '\\', 'u',
+                xdigits[(lo >> (4 * 3)) & 0xF],
+                xdigits[(lo >> (4 * 2)) & 0xF],
+                xdigits[(lo >> (4 * 1)) & 0xF],
+                xdigits[(lo >> (4 * 0)) & 0xF],
+            }), (sizeof("\\u1234") - 1) * 2);
+          }
+          break;
+        }
+      }
+      i += shift;
+    }
+    ga_append(gap, '"');
+    xfree(tofree);
+  }
+  return OK;
+}
+
+#undef CONV_STRING
+#define CONV_STRING(buf, len) \
+    do { \
+      if (convert_to_json_string(gap, (const char *) (buf), (len)) != OK) { \
```

Comment:
Some of the `CONV_STRING` macros are used with `STRLEN(buf)` as length argument (line 309). Can we be sure that `buf != NULL`? `convert_to_json_string()` has a check for `buf == NULL`.